src/main/java/com/yelp/nrtsearch/server/luceneserver/search/SortParser.java

```diff
@@ -0,0 +1,142 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.yelp.nrtsearch.server.luceneserver.search;
+
+import com.yelp.nrtsearch.server.grpc.SearchResponse;
+import com.yelp.nrtsearch.server.grpc.SortType;
+import com.yelp.nrtsearch.server.luceneserver.SearchHandler;
+import com.yelp.nrtsearch.server.luceneserver.field.FieldDef;
+import com.yelp.nrtsearch.server.luceneserver.field.properties.Sortable;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import org.apache.lucene.search.FieldDoc;
+import org.apache.lucene.search.Sort;
+import org.apache.lucene.search.SortField;
+
+/**
+ * Class to handle creation of a {@link Sort} used to sort documents by field values for queries.
+ */
+public class SortParser {
+
+  private static final String DOCID = "docid";
+  private static final String SCORE = "score";
+
+  private SortParser() {}
+
+  /**
+   * Decodes a list of request {@link SortType} into the corresponding {@link Sort}.
+   *
+   * @param fields list of {@link SortType} from grpc request
+   * @param sortFieldNames mutable list which will have all sort field names added in sort order,
+   *     may be null
+   * @param queryFields collection of all possible fields which may be used to sort
+   */
+  public static Sort parseSort(
+      List<SortType> fields, List<String> sortFieldNames, Map<String, FieldDef> queryFields)
+      throws SearchHandler.SearchHandlerException {
+    List<SortField> sortFields = new ArrayList<>();
+    for (SortType sub : fields) {
+      String fieldName = sub.getFieldName();
+      SortField sf;
+      if (sortFieldNames != null) {
+        sortFieldNames.add(fieldName);
+      }
+      if (fieldName.equals(DOCID)) {
+        sf = SortField.FIELD_DOC;
+      } else if (fieldName.equals(SCORE)) {
+        sf = SortField.FIELD_SCORE;
+      } else {
+        FieldDef fd = queryFields.get(fieldName);
+        if (fd == null) {
+          throw new SearchHandler.SearchHandlerException(
+              String.format(
+                  "field: %s was not registered and was not specified as a virtualField",
+                  fieldName));
+        }
+
+        if (!(fd instanceof Sortable)) {
+          throw new IllegalArgumentException(
+              String.format("field: %s does not support sorting", fieldName));
+        }
+
+        sf = ((Sortable) fd).getSortField(sub);
+      }
+      sortFields.add(sf);
+    }
+
+    return new Sort(sortFields.toArray(new SortField[0]));
```

Comment:
Should we  allow the SortField to take in a custom `FieldComparatorSource`. 

```
/** Creates a sort, possibly in reverse, with a custom comparison function.
   * @param field Name of field to sort by; cannot be <code>null</code>.
   * @param comparator Returns a comparator for sorting hits.
   * @param reverse True if natural order should be reversed.
   */
  public SortField(String field, FieldComparatorSource comparator, boolean reverse) {
    initFieldType(field, Type.CUSTOM);
    this.reverse = reverse;
    this.comparatorSource = comparator;
  }
```

Not entirely sure how this would work on our custom scoring scripts since for virutal fields this seems to create a `FieldComparator.DoubleComparator`. Wondering if might be good to have our own custom comparator. Probably add a test for now to show how this might work on Script/virtual fields?